### PR TITLE
OAK-9265 : Recovery utility function of LastRevRecoveryAgent

### PR DESCRIPTION
**Pull request by Vinod-holani**
should not process previous docs

### CODE REVIEW

**Comment by mreutegg** [comment]
Alternatively the implementation(s) of MissingLastRevSeeker could be changed to filter our split documents. But apply the filtering later also works.

Would it be possible to have a test that reproduces the problem?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/LastRevRecoveryAgent.java [hunk 1]
```diff
@@ -310,6 +310,12 @@ public void sweepUpdate(Map<Path, UpdateOp> updates)
         long lastLog = startOfScan;
 
         for (NodeDocument doc : suspects) {
+        	
+        	//Should skip previous documents
+        	if(!doc.getPath().isAbsolute()) {
```
Comment (mreutegg) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/LastRevRecoveryAgent.java]: I would rather make a decision on NodeDocument.getSplitDocType(). WDYT?

Comment (Vinod-holani) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/LastRevRecoveryAgent.java]: Yes that is better, I will use NodeDocument.isSplitDocument() to check previous/split document

**Comment by Vinod-holani** [comment]
> Would it be possible to have a test that reproduces the problem?

I already thought about this but recover function only returns no. of inconsistent documents, in logging the scanned document no. will be different but not sure should we try this. Could you suggest something for test case? 

And in current code infinite loop issue won't occur due to Path class implementation, so original issue won't be reproducible.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/LastRevRecoveryAgent.java (hunk 1 above)
Comment (mreutegg) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/LastRevRecoveryAgent.java]: This is even better. +1

**Comment by mreutegg** [comment]
Hmm, it is indeed difficult to come up with a test that fails with the current code. What about a test that runs just fine in trunk, but would fail in the maintenance branches without the Path class? Though, I don't see how to write such a test either. The infinite loop would simply cause the test to run forever.

**Comment by Vinod-holani** [comment]
@mreutegg @reschke could you please review this PR?

**Comment by reschke** [comment]
sorry, currently not working on Oak...

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java [hunk 2]
```diff
@@ -216,4 +220,26 @@ public void getClusterNodeInfo() {
...
+            b1 = dns.getRoot().builder();
+            b1.child(nodeName).setProperty("prop",i);
+            dns.merge(b1, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+        }
+        dns.runBackgroundOperations();
+        int docs = 0;
+        //seeker should return only non split documents
+        for(Document doc : seeker.getCandidates(0)) {
+        	docs++;
+        }
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Can be simplified with:
```suggestion
        //seeker should return only non split documents
        int docs = Iterables.size(seeker.getCandidates(0));
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java [hunk 3]
```diff
@@ -54,6 +55,18 @@ public BasicDocumentStoreTest(DocumentStoreFixture dsf) {
...
+        Revision.resetClockToDefault();
+        markDocumentsForCleanup();
+    }
+
+    private void markDocumentsForCleanup() {
+        for (NodeDocument doc : Utils.getAllDocuments(ds)) {
+            removeMe.add(doc.getId());
+        }
+    }
+
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java]: Why is this needed in BasicDocumentStoreTest? AFAICS none of the tests uses a Clock. I think this should rather go to `MissingLastRevSeekerTest` and clean up documents created by the new `getNonSplitDocs()` test.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java [hunk 4]
```diff
@@ -216,4 +220,26 @@ public void getClusterNodeInfo() {
...
+        assertEquals(2, docs);
+    }
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:256]: The DocumentNodeStore is not disposed.
```suggestion
        dns.dispose();
    }
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java [hunk 5]
```diff
@@ -216,4 +220,26 @@ public void getClusterNodeInfo() {
...
+        NodeBuilder b1 = dns.getRoot().builder();
+        b1.child(nodeName);
+        dns.merge(b1, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+        //Modify and commit changes on this node 100 times to create a split document
+        for (int i = 0; i < NUM_REVS_THRESHOLD; i++) {
+            b1 = dns.getRoot().builder();
+            b1.child(nodeName).setProperty("prop",i);
+            dns.merge(b1, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+        }
+        dns.runBackgroundOperations();
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:250]: When triggering background operations manually, the background threads should be disabled on the DocumentNodeStore with `DocumentNodeStoreBuilder.setAsyncDelay(0)`.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java [hunk 6]
```diff
@@ -216,4 +220,26 @@ public void getClusterNodeInfo() {
     private ClusterNodeInfoDocument getClusterNodeInfo(int clusterId) {
         return seeker.getClusterNodeInfo(clusterId);
     }
+    
+    @Test
+    public void getNonSplitDocs() throws Exception {
+        String nodeName = this.getClass().getName() + "-foo";
+        DocumentNodeStore dns = getBuilder().setDocumentStore(store).getNodeStore();
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: The DocumentNodeStore should use the clock setup by this test. Otherwise it can be unstable because `Revision` and `ClusterNodeInfo` are using the virtual clock. Unfortunately it is also necessary to wrap the store and prevent disposal of the underlying store.
```suggestion
        DocumentNodeStore dns = getBuilder().clock(clock).setAsyncDelay(0)
                .setDocumentStore(new DocumentStoreWrapper(store) {
                    @Override
                    public void dispose() {
                        // do not close underlying store, otherwise cleanup
                        // cannot remove documents after the test
                    }
                }).getNodeStore();
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 5 above)
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:250]: At this point the test should remember the documents for the cleanup.
```suggestion
        dns.runBackgroundOperations();
        // remember all documents for cleanup
        for (NodeDocument doc : Utils.getAllDocuments(ds)) {
            removeMe.add(doc.getId());
        }
```

**Review by mreutegg — CHANGES_REQUESTED** [review]

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java (hunk 3 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java]: There was one node created by this class which was not getting cleaned up and causing the newly created test to fail. Thats why I added this cleanup over here. I will dig deeper in this.

Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java]: Basically this [0] is the culprit function creating a node but not able to delete it. Need to either fix this or can add after cleanup function as already done.

[0] : https://github.com/apache/jackrabbit-oak/blob/trunk/oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java#L1144

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 6 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Does this wrapping is required if we are not disposing the nodestore?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java (hunk 3 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BasicDocumentStoreTest.java]: Fixed the culprit test case from BasicDocumentStoreTest class

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 5 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:250]: Done

Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:250]: Done

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 2 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Done

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 4 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:256]: As per above review comment we should not do this right?

Comment (mreutegg) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:256]: Actually, the test _should_ dispose the `DocumentNodeStore` when it is not needed anymore. Some tests use a `DocumentMKBuilderProvider` rule that takes care of disposing the node store automatically after the test, but this test creates the node store manually.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 6 above)
Comment (mreutegg) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Disposal is needed, otherwise some background threads continue to run and occupy a clusterId. E.g. the lease update thread.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 4 above)
Comment (Vinod-holani) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java:256]: Ok so we do need to dispose the DocumentNodeStore but not underlying DocumentStore for this test case. I was confused with these two comments.
